### PR TITLE
cargo-pgx: deny clippy::perf

### PR DESCRIPTION
**Pull request by workingjubilee**
this catches the lints
- clippy::expect_fun_call
- clippy::single_char_pattern
- clippy::needless_clone
- clippy::or_fun_call

### CODE REVIEW

**Review by BradyBonnette — APPROVED** [review]
Thanks for the feedback, learned yet more things!

**Comment by BradyBonnette** [comment]
@workingjubilee well I have no idea how this happened, but I meant to approve https://github.com/tcdi/pgx/pull/930 not this one. Sorry about that!

File: cargo-pgx/src/command/install.rs [hunk 1]
```diff
@@ -101,7 +101,7 @@ pub(crate) fn install_extension(
     base_directory: Option<PathBuf>,
     features: &clap_cargo::Features,
 ) -> eyre::Result<()> {
-    let base_directory = base_directory.unwrap_or("/".into());
+    let base_directory = base_directory.unwrap_or_else(|| PathBuf::from("/"));
```
Comment (workingjubilee) [inline cargo-pgx/src/command/install.rs:104]: Note that this allocates, so it doesn't break the "actually `clippy::or_fun_call` is wrong here" criteria I mentioned before. Unsure if it matters for perf tho'. However, I think this lint should be allowed, after this.

File: cargo-pgx/src/command/install.rs [hunk 2]
```diff
@@ -393,7 +393,7 @@ pub(crate) fn get_version(manifest_path: impl AsRef<Path>) -> eyre::Result<Strin
                     .wrap_err("Couldn't parse manifest")?;
 
                 let version = manifest.package.ok_or(eyre!("no `[package]` section found"))?.version;
-                Ok(version.to_string())
+                Ok(version)
```
Comment (workingjubilee) [inline cargo-pgx/src/command/install.rs:396]: Explicit extra allocs are things rustc/LLVM struggle to elide, so this is a good one to get.

File: cargo-pgx/src/command/schema.rs [hunk 3]
```diff
@@ -429,23 +429,23 @@ pub(crate) fn generate_schema(
             unsafe extern "Rust" fn() -> pgx_sql_entity_graph::RustToSqlMapping,
         > = lib
             .get("__pgx_sql_mappings".as_bytes())
-            .expect(&format!("Couldn't call __pgx_sql_mappings"));
+            .expect("Couldn't call __pgx_sql_mappings");
```
Comment (workingjubilee) [inline cargo-pgx/src/command/schema.rs:432]: Explicit extra alloc.

File: cargo-pgx/src/command/schema.rs [hunk 4]
```diff
@@ -429,23 +429,23 @@ pub(crate) fn generate_schema(
...
         sql_mapping = sql_mappings_symbol();
 
         let symbol: libloading::os::unix::Symbol<
             unsafe extern "Rust" fn() -> eyre::Result<pgx_sql_entity_graph::ControlFile>,
         > = lib
             .get("__pgx_marker".as_bytes())
-            .expect(&format!("Couldn't call __pgx_marker"));
+            .expect("Couldn't call __pgx_marker");
```
Comment (workingjubilee) [inline cargo-pgx/src/command/schema.rs:439]: Explicit extra alloc.

File: cargo-pgx/src/command/schema.rs [hunk 5]
```diff
@@ -555,7 +555,7 @@ fn create_stub(
     so_rustc_invocation.stderr(Stdio::inherit());
 
     if let Some(rustc_flags_str) = std::env::var("RUSTFLAGS").ok() {
-        let rustc_flags = rustc_flags_str.split(" ").collect::<Vec<_>>();
+        let rustc_flags = rustc_flags_str.split(' ').collect::<Vec<_>>();
```
Comment (workingjubilee) [inline cargo-pgx/src/command/schema.rs:558]: I feel doubtful that this isn't monomorphized away by LLVM, but I've noticed subtle differences in these cases, so very possible still.

File: cargo-pgx/src/command/schema.rs [hunk 6]
```diff
@@ -429,23 +429,23 @@ pub(crate) fn generate_schema(
...
             symbol().expect("Failed to get control file information"),
         );
         entities.push(control_file_entity);
 
         for symbol_to_call in fns_to_call {
             let symbol: libloading::os::unix::Symbol<unsafe extern "Rust" fn() -> pgx_sql_entity_graph::SqlGraphEntity> =
-                lib.get(symbol_to_call.as_bytes())
-                    .expect(&format!("Couldn't call {:#?}", symbol_to_call));
+                lib.get(symbol_to_call.as_bytes()).unwrap_or_else(|e|
+                    panic!("Couldn't call {:#?}", symbol_to_call));
```
Comment (workingjubilee) [inline cargo-pgx/src/command/schema.rs:448]: Pushes alloc and everything into the `unwrap_or_else` case, so this is good.

File: cargo-pgx/src/command/schema.rs [hunk 7]
```diff
@@ -320,7 +320,7 @@ pub(crate) fn generate_schema(
 
     let so_extension = if cfg!(target_os = "macos") { ".dylib" } else { ".so" };
 
-    lib_so.push(&format!("lib{}{}", package_name.replace("-", "_"), so_extension));
+    lib_so.push(&format!("lib{}{}", package_name.replace('-', "_"), so_extension));
```
Comment (workingjubilee) [inline cargo-pgx/src/command/schema.rs:323]: I really doubt this one improves anything lol.

**Comment by workingjubilee** [comment]
Heads up for thom since IIRC he was tinkering inside cargo-pgx.

**Pull request merged**